R/calculate-grades.R

```diff
@@ -1,261 +1,441 @@
-#' Aggregate assignment scores
+#' Calculate Grades
+#' This function calculates all grades based on the policy file.
+#'
+#' @param gs A processed data in wide format 
+#' @param policy A policy file
+#'
+#' @return A data frame
+#'
+#' @importFrom dplyr select
+#' 
+#' @export
+
+get_grades <- function(gs, policy){
+  
+  # flatten policy file
+  policy <- flatten_policy(policy)
+  
+  # convert gs into a matrix with only assignment info
+  grades_mat <- gs |>
+    #only assignment info
+    select(-get_id_cols(gs)) |> 
+    #convert lateness into minutes
+    mutate_at(vars(ends_with(" - Lateness (H:M:S)")), convert_to_min) |> 
+    #convert to matrix
+    data.matrix()
+  grades_mat[is.na(grades_mat)] <- 0
+  #rownames are SID of student
+  rownames(grades_mat) <- gs$SID
+  #pre_allotted cols
+  categories <- unlist(map(policy$categories, "category"))
+  empty <- matrix(nrow = length(gs$SID), ncol = length(categories)*3)
+  colnames(empty) <- paste0(rep(categories, each = 3), c("", " - Max Points", " - Lateness (H:M:S)"))
+  grades_mat <- cbind(grades_mat, empty)
+  
+  
+  #iterate through each policy item
+  for (policy_item in policy$categories){
+    #compute calculations for each policy file and save into grades matrix
+    grades_mat <- get_category_grade(grades_mat, policy_item)
+  }
+  
+  grades <- grades_mat |>
+    data.frame()
+  names(grades) <- colnames(grades_mat)
+  grades$SID <- rownames(grades_mat)
+  #add back ID cols
+  
+  return (grades)
+}
+
+#' Calculate Single Category Grade
+#' This function calculates all grades based on the policy file.
+#'
+#' @param grades_mat Matrix with assignments + associated cols for that category
+#' @param policy_item An item from policy file
+#'
+#' @return A data frame
+#'
+#' 
+#' @export
+
+get_category_grade <- function(grades_mat, policy_item){
+  #get all keys except category and assignments (which are not functions)
+  keys <- names(policy_item)[-which(names(policy_item) %in% c("category", "assignments", "weights"))]
+  #all assignments + their associated cols for this category
+  for (key in keys){
+    grades_mat <- get(key)(grades_mat, policy_item[[key]], policy_item$category, 
+                           policy_item$assignments, policy_item$weights)
+  }
+  
+  ## insert code to re-add new/editted columns from grades_mat into grades_mat
```

Comment:
what is this comment for?

Reply:
That was a note to myself that isn't relevant anymore because I changed the approach